crates/symbolicator/src/services/symbolication.rs

```diff
@@ -371,40 +314,91 @@ impl ModuleListBuilder {
         };
 
-        let (info, marked) = &mut self.inner[info_index];
+        let (info, _) = &self.inner[info_idx];
         let HexValue(image_addr) = info.raw.image_addr;
-        let should_mark = match info.raw.image_size {
+        let includes_addr = match info.raw.image_size {
             Some(size) => addr < image_addr + size,
             // If there is no image size, the image implicitly counts up to the next image. Because
             // we know that the search address is somewhere in this range, we can mark it.
             None => true,
         };
 
-        if should_mark {
-            *marked = true;
+        if includes_addr {
+            Some(info_idx)
+        } else {
+            None
         }
     }
 
+    fn process_stacktraces(&mut self, stacktraces: &[RawStacktrace]) {
+        for trace in stacktraces {
+            if let Some(first_frame) = trace.frames.first() {
+                let addr = first_frame.instruction_addr.0;
+                self.mark_referenced(addr);
+            }
+            for frame_pair in trace.frames.windows(2) {
+                if let [prev_frame, next_frame] = frame_pair {
+                    let addr = next_frame.instruction_addr.0;
+                    self.mark_referenced(addr);
+
+                    if let Some(info_index) = self.find_module_index(prev_frame.instruction_addr.0)
+                    {
+                        let (info, _) = &mut self.inner[info_index];
+                        if info.unwind_status.is_none() {
+                            // We report this error once per missing module in a minidump.
+                            sentry::capture_message(
+                                "Module marked as used but not found",
+                                sentry::Level::Error,
+                            );
+                            info.unwind_status = Some(ObjectFileStatus::Missing);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /// Marks the module loaded at the given address as referenced.
+    ///
+    /// The respective module will always be included in the final list of modules.
+    pub fn mark_referenced(&mut self, addr: u64) {
+        let info_index = match self.find_module_index(addr) {
+            Some(idx) => idx,
+            None => return,
+        };
+
+        let (_, marked) = &mut self.inner[info_index];
+        *marked = true;
+    }
+
     /// Returns the modules list to be used in the symbolication response.
     pub fn build(self) -> Vec<CompleteObjectInfo> {
         self.inner
             .into_iter()
-            .filter(|(info, marked)| *marked || info.raw.debug_id.is_some())
-            .map(|(info, _)| info)
+            .filter(|(info, marked)| {
+                metric!(
+                    counter("symbolication.unwind_status") += 1,
+                    "status" => info.unwind_status.unwrap_or(ObjectFileStatus::Unused).name(),
```

Comment:
`unwrap_or_default()` also works as Unused is the default value.  And if you also put the `.map()` before the `.filter()` that is probably a more succinct  way of expressing this.